scripts/mod_loader/modapi/palette.lua

```diff
@@ -0,0 +1,354 @@
+				local index = pawnType.ImageOffset + 1
+				
+				if
+					palettes:getHash(index) ~= nil         and
+					palettes:getPaletteImage(index) == nil
+				then
+					
+					local animation = ANIMS[pawnType.Image .."_ns"]
+					if
+						type(animation) == 'table'        and
+						type(animation.Image) == 'string'
+					then
+						palettes:setPaletteImage(index, animation.Image)
+					end
+				end
+			end
+		end
+		
+		local new_paletteOrder = {}
+		for i = 2, 11 do
+			new_paletteOrder[i] = palettes:getHash(i)
+		end
+		
+		config.paletteOrder = new_paletteOrder
+	end)
+end
+
+local function migrateColorMaps()
+	local fromIndex = palettes:getCount() + 1
+	local toIndex = GetColorCount()
+	
+	if fromIndex > toIndex then return end
+	
+	for i = fromIndex, toIndex do
+		local palette = GetColorMap(i)
+		local hash = getVanillaId(i) or getFurlId(i) or getPaletteHash(palette)
+		
+		palettes:addPalette(hash, palette)
+	end
+	
+	GetColorMap = modApi.getColorMap
+	GetColorCount = modApi.getColorCount
+end
+
+function modApi:finalizePalettes()
+	
+	migrateColorMaps()
+	saveDefaultPaletteOrder()
+	loadPaletteOrder()
+	
+	local playerAnimationHeight = GetColorCount()
+	
+	for anim_name, animation in pairs(ANIMS) do
+		if
+			type(animation) == 'table'         and
+			type(animation.Height) == 'number' and
+			type(animation.Image) == 'string'
+		then
+			local isPlayerUnitAnimation = self:stringStartsWith(animation.Image, "units/player")
+			
+			if isPlayerUnitAnimation then
+				animation.Height = playerAnimationHeight
+			end
+		end
+	end
+	
+    -- update base objects that cannot be determined
+	-- with Image starting with "units/player"
+    ANIMS.MechColors = playerAnimationHeight
+    ANIMS.MechIcon.Height = playerAnimationHeight
+	
+    ANIMS.BaseUnit.Height = 1
+    ANIMS.EnemyUnit.Height = 3
+	
+	-- remove ability to add palettes and
+	-- clear these functions for further use.
+	canAddPalettes = false
+	self.finalizePalettes = nil
+end
+
+function modApi:addPalette(palette, id)
+	Assert.True(canAddPalettes, "Cannot add palettes after game init")
+	Assert.Equals('table', type(palette), "Argument #1")
+	Assert.Equals('string', type(id), "Argument #2")
+	
+	migrateColorMaps()
+	
+	if palettes:getIndex(id) ~= nil then
+		LOGF("A palette with id %s has already been added", id)
+		return
+	end
+	
+	local new_palette = {}
+	for color_name, color in pairs(palette) do
+		local color_index = COLOR_NAME_2_INDEX[color_name]
+		
+		if color_index then
+			if type(color) == 'table' then
+				for i = 1, 3 do
+					Assert.Equals('number', type(color[i]), string.format("Argument #1 - color %s entry %s", color_name, i))
+				end
+				
+				color = GL_Color(color[1], color[2], color[3])
+			end
+			
+			new_palette[color_index] = color
+		end
+	end
+	
+	for i = 1, 8 do
+		Assert.Equals('userdata', type(new_palette[i]), string.format("Argument #1 - incomplete palette"))
+	end
+	
+	palettes:addPalette(id, new_palette)
+end
+
+function modApi:getPaletteIndex(id)
```

Comment:
I'm not sure I like calling this `getPaletteIndex`, since that typically means the true index. This function is closer to `getPaletteImageOffset`, makes it clear it uses a different indexing scheme.

Reply:
I think I can get behind that.